### ISSUE TITLE
bug: to_parquet() writes incorrect data without previous .cache()

### ISSUE THREAD

**Issue by NickCrews** [ibis-project/ibis#6091]
### What happened?

OK, I don't have a nice repro for this because I'm not sure exactly what is causing it, and it's too complicated for me to share it all. If you could help me figure out where exactly to look then I gladly will.

I am loading some data, doing a bunch of processing, and then writing `.to_parquet()`. It writes fine, but several columns are always null, when they should have contents. If I call .cache() before writing, then they write with the correct contents.

Here is the output of `ibis.show_sql(table)` just before the `.to_parquet()` if that is helpful: [log.sql](https://github.com/ibis-project/ibis/files/11327266/log.txt)

When I do this, I get these warnings, if you think that could be related. Using `sqlglot==11.5.7`
```
WARNING:sqlglot:REGEXP_EXTRACT does not support arg(s) ['position']
WARNING:sqlglot:REGEXP_EXTRACT does not support arg(s) ['position']
WARNING:sqlglot:REGEXP_EXTRACT does not support arg(s) ['position']
WARNING:sqlglot:REGEXP_EXTRACT does not support arg(s) ['position']
WARNING:sqlglot:REGEXP_EXTRACT does not support arg(s) ['position']
WARNING:sqlglot:REGEXP_EXTRACT does not support arg(s) ['position']
WARNING:sqlglot:REGEXP_EXTRACT does not support arg(s) ['position']
```

Let me know how you think I can debug this further.

### What version of ibis are you using?

5.1.0

### What backend(s) are you using, if any?

duckdb

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by cpcloud** [comment]
Another thing that would be useful is knowing the schema of the input CSV, i.e. the types of these columns:

```
   3   │     t5."Date" AS "Date",
   4   │     t5."Payment Type" AS "Payment Type",
   5   │     t5."Amount" AS "Amount",
   6   │     t5."Last/Business Name" AS "Last/Business Name",
   7   │     t5."First Name" AS "First Name",
   8   │     t5."Address" AS "Address",
   9   │     t5."City" AS "City",
  10   │     t5."State" AS "State",
  11   │     t5."Zip" AS "Zip",
  12   │     t5."Country" AS "Country",
  13   │     t5."Occupation" AS "Occupation",
  14   │     t5."Employer" AS "Employer",
  15   │     t5."Election Name" AS "Election Name",
  16   │     t5."Municipality" AS "Municipality",
  17   │     t5."Office" AS "Office",
  18   │     t5."Filer Type" AS "Filer Type",
  19   │     t5."Name" AS "Name",
  20   │     t5."Report Year" AS "Report Year",
  21   │     t5."Submitted" AS "Submitted"
```

**Comment by jcrist** [comment]
Hi @NickCrews, can you provide some more context here ^^?

**Comment by NickCrews** [comment]
Yeah sorry, I realize this is in my court, I just haven't found time to dig back into this.

**Comment by cpcloud** [comment]
I think we need the schema from the columns in my previous comment to be able to try to generate some fake data, so that we can run the SQL you posted against the fake data and try to reproduce.

**Comment by NickCrews** [comment]
I'm on vacation until next Tuesday, I'll take a look next week. Thanks!

**Comment by NickCrews** [comment]
OK, my code has evolved so much that I can't get the exact same error, but I get something similar, in the sense that things work with .cache(), and fail without cache(). This time, I get an error about not being able to find a memtable.

I'm still using ibis 5.1.0, but duckdb 0.8.0. I can update to ibis 6.0.0 if you think that's significant.

If I try to write to parquet without a .cache(), I get
```
sqlalchemy.exc.ProgrammingError: (duckdb.CatalogException) Catalog Error: Table with name _ibis_pyarrow_memtable_3vsypavcsr7399y9a3cwbk8bt does not exist!
Did you mean "pg_tablespace"?
```
from the attached SQL.
[without_cache.txt](https://github.com/ibis-project/ibis/files/12045466/err_sql.txt)

I still don't know how to show the SQL of the DDL statement that happens within the .cache(), as I assume that is going to be different from the above sql.

**Comment by gforsyth** [comment]
Hey @NickCrews -- I have a fix incoming for this (or at least, I have a reproducer for that error that I can fix).

**Comment by NickCrews** [comment]
Thanks @gforsyth , that PR looks great for my bug 2. I think you should just close out this issue once that PR is merged since I'm not able to help you get a repro for bug 1. If that comes up again for me, I'll just file a new issue. Thank you!

**Comment by gforsyth** [comment]
> If that comes up again for me, I'll just file a new issue. Thank you!

Sounds good!  Thanks @NickCrews and :crossed_fingers: that we've seen the last of that one.
